### ISSUE TITLE
Survey of SNMP Libraries for Java

### ISSUE THREAD

**Issue by janiree** [research-iobserve/snmp-monitoring#1]
- Java Dynamic Management Kit & OpenDMK
   - Description: Inbuilt Java Classes and tools for creating monitoring Solutions based on the SNMP Standard
   - Develloper: Oracle (Sun Microsystems) (archived project)
   - Links: https://javaee.github.io/other-migrated-projects.html, http://dev.cs.ovgu.de/java/jdmk/, http://dev.cs.ovgu.de/java/jdmk/api/
   - Notes: no support since 2007, comatose project
 - SNMP4J
   - Description: Open Source SNMP Implementation for Java
   - Develloper: Frank Fock
   - Links: http://www.snmp4j.org/, http://www.snmp4j.org/doc/index.html
   - Notes: noted by many sources to be the go-to SNMP solution currently, active forum and mailing list
 - netsnmpj
   - Description: Open Source Library to perform SNMP operations across all protocols
   - Develloper: Andrew E.
   - Links: http://netsnmpj.sourceforge.net/
   - Notes: restricted to Linux
 - joesnmp
   - Description: open sourde Java SNMP library
   - Develloper: OpenNMS.org
   - Links: https://sourceforge.net/projects/joesnmp/
   - Notes: long time since the last update
 - Westhawk's Java SNMP stack
   - Description: Lightweight Java SNMP Implementation, capable of sending and recieving, limited Agent functionality
   - Develloper: Westhawk Ltd.
   - Links: https://sourceforge.net/projects/westhawksnmp/
   - Notes: still being maintained
 - JSNMP
   - Description: SNMP project with full support for SNMPv1, v2 and v3
   - Develloper: jSNMP Enterprise
   - Links: http://www.jsnmp.com/products.html
   - Notes: project has been discontinued, unsupported license available for purchase
 - Network Ferret
   - Description: Designem for OEM, Java based network discovery and topology engine
   - Develloper: AMT
   - Links: http://www.altmantech.com/features.htm
   - Notes: not free, not open source
 - WebNMS SNMP API & WebNMS SNMP Agent Toolkit Java
   - Description: development toolkit for SNMP-based network management applications
   - Develloper: ZOHO Corp.
   - Links: https://www.webnms.com/snmp/index.html, http://www.webnms.com/javaagent/index.html
   - Notes: only trial licenses are free(45 days), register to download
 - iReasoning SNMP library & iReasoning SNMP agent builder
   - Description: library for building SNMP network management applications
   - Develloper: iReasoning Inc.
   - Links: http://www.ireasoning.com/snmpapi.shtml, http://www.ireasoning.com/snmpagent.shtml
   - Notes: only trial versions free (30 days)
 - WIN-SNMP
   - Description:  API for developing SNMPv1, SNMPv2c and SNMPv3 SNMP manager and agent applications
   - Develloper: MG-SOFT Corporation
   - Links: http://www.mg-soft.si/SNMP-Lab.html
   - Notes: primarily aimed at MS-Windows, though linux versions are also provided, only trial versions available for free (30 days)
 - DynamicSNMP® Java SNMP
   - Description: Java libraries to implement SNMP managers and agents
   - Develloper: Monfox, LLC
   - Links: http://www.monfox.com/dsnmp/index.html
   - Notes: only trial versions for free (30 days), register to download
 - IP*Works! S/SNMP
   - Description: toolkit for building SNMP Agent and manager applications
   - Develloper: /n software inc.
   - Links: https://www.nsoftware.com/ipworks/snmp/
   - Notes: only trial versions free (30 days),

**Comment by rju** [comment]
That looks promising. Unfortunately, there are quite a few commercial libraries, which are only partially usable due to licensing issues. Therefore, you should focus on the open-source variants. Start with:
* Westhawk's Java SNMP stack (I do not know that, but maybe this is good enough to realize a SNMP reader)
* SNMP4J (as it is still active developed, also there are good example on the Internet)
Secondaries are:
* JSNMP
* joesnmp
For those you should assess how complicated their code base is, as we might need to fork their project.
Third level are:
* netsnmpj
Only, if you cannot create a working solution with the libraries listed above, you can inspect the commercial libraries. However, this would greatly limit its usability. Therefore, we should avoid them.

**Comment by rju** [comment]
After looking at the code of Westhawk's Java SNMP stack, they seem to provide a set of examples, which might be helpful to test things extensively. Also we could setup an SNMP server on a Linux box so you ca query it.

**Comment by janiree** [comment]
Yes, that was my line of thought as well. I would be quite surprised if a working solution would require proprietary tools. Next on my agenda is trying out Westhawk's and SNMP4J. Unfortunately i seem to have made a mistake with JSNMP, which is in fact not open source.
The Secondary options are thus joesnmp and netsnmpj, which i will also be looking at but i think the first two show the most promise so far.
I will also be looking at openJDMK, if only for reference.
